File: tradeit_bot.py
import logging
import re

from selenium.webdriver import Firefox
from selenium.common.exceptions import NoSuchElementException
from selenium.webdriver.common.by import By
from selenium.webdriver.remote.webelement import WebElement
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait


class TradeitBot:
    URL = 'https://tradeit.gg/'
    ITEMS_XPATH = '//div[@id="sinv-loader"]//li'

    def __init__(self, driver: Firefox):
        self.wait_long = WebDriverWait(driver, 30)
        self.wait_short = WebDriverWait(driver, 2)
        self.__driver = driver
        self.__driver.get(self.URL)
        self.wait_long.until(EC.presence_of_element_located((By.ID, 'trading-steps')))

    def search(self, item_name: str, game: str):
        game_options = self.__driver.find_element_by_xpath(f'//div[@id="bot-inv-column"]//li[@game="{game}"]')
        if 'sactive-game' not in game_options.get_attribute('class'):
            self.wait_long.until(
                EC.element_to_be_clickable((By.XPATH, f'//div[@id="bot-inv-column"]//li[@game="{game}"]'))).click()
            self.__wait_for_items()

        search_input = self.__driver.find_element_by_id('ssearch')
        if search_input.text != item_name:
            bot_items = self.__driver.find_elements_by_xpath(self.ITEMS_XPATH)
            last_item_html = bot_items[-1].get_attribute('innerHTML') if bot_items else ''
            search_input.send_keys(item_name)
            self.__wait_for_refreshed_items(last_item_html)

        updated = False
        bot_items = self.__driver.find_elements_by_xpath(self.ITEMS_XPATH)
        last_item_html = bot_items[-1].get_attribute('innerHTML') if bot_items else ''
        for bot_item in bot_items:
            try:
                multiplier = bot_item.find_element_by_xpath('.//div[@class="multiplier"]')
                self.wait_short.until(
                    lambda _: multiplier if multiplier.is_displayed() and multiplier.is_enabled() else False).click()
                updated = True
            except NoSuchElementException:
                pass

        self.__wait_for_refreshed_items(last_item_html) if updated else None

        items = map(self.__web_element_to_dict, self.__driver.find_elements_by_xpath(self.ITEMS_XPATH))
        search_input.clear()
        return items

    def __wait_for_items(self):
        def wait_function(web_driver: Firefox):
            text = web_driver.find_element_by_id('sinv-empty')
            elements = web_driver.find_elements_by_xpath(self.ITEMS_XPATH)
            # logging.debug(f"Waiting for items - Message is displayed: {text.is_displayed()} - Items size: {len(elements)}")
            return text.is_displayed() or elements

        self.wait_long.until(wait_function)

    def __wait_for_refreshed_items(self, initial_state: str):
        def wait_function(web_driver: Firefox):
            bot_items = web_driver.find_elements_by_xpath(self.ITEMS_XPATH)
            current_state = bot_items[-1].get_attribute('innerHTML') if bot_items else ''
            equals = current_state == initial_state
            # logging.debug(f"Waiting for updated items - Initial state lenght: {len(initial_state)} - Current state lenght: {len(current_state)}")
            return not equals

        self.wait_short.until(wait_function)

    def __web_element_to_dict(self, element: WebElement) -> dict:
        float_locator = By.XPATH, './/span[contains(text(), "Float")]//parent::div'
        paint_locator = By.XPATH, './/span[contains(text(), "Paint")]//parent::div'
        price_locator = By.XPATH, './/span[@class="pricetext"]'
        wear_locator = By.XPATH, './/div[@class="quality"]'
        stattrack_locator = By.XPATH, './/div[@class="stattrak"]'
        locked_locator = By.XPATH, './/div[@class="bot-icon"]'

        return {
            'name': str(element.get_attribute('data-original-title')).strip(),
            'game': str(element.get_attribute('data-appid')).strip(),
            'floatvalue': self.__value_of(self.__get_text(element, float_locator)),
            'paint': self.__value_of(self.__get_text(element, paint_locator)),
            'price': self.__value_of(self.__get_text(element, price_locator)),
            'wear': self.__get_text(element, wear_locator),
            'stattrack': True if self.__get_text(element, stattrack_locator) else False,
            'locked': self.__get_text(element, locked_locator)
        }

    @staticmethod
    def __get_text(element: WebElement, locator: tuple):
        try:
            parent = element.find_element(*locator)
            text = str(parent.get_attribute('textContent')).strip()
            children = parent.find_elements(By.XPATH, './*')
            for child in children:
                text = text.replace(child.get_attribute('textContent'), '', 1).strip()
            return text
        except NoSuchElementException:
            return None

    @staticmethod
    def __value_of(text: str):
        text = re.sub(r'[^\d.]', '', text) if text else ''
        try:
            value = float(text)
        except ValueError:
            return None
        else:
            return int(value) if value.is_integer() else value


if __name__ == '__main__':
    logging.basicConfig(level=logging.INFO)

    driver = Firefox(executable_path='')
    bot = TradeitBot(driver)
    items = bot.search('AK-47 | Elite Build', '730')

    for item in items:
        print(item)

File: TradeitBot.java
import lombok.extern.apachecommons.CommonsLog;
import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;


@CommonsLog
public class TradeitBot {

    public static final String URL = "https://tradeit.gg/";
    private static final By ITEMS_LOCATOR = By.xpath("//div[@id='sinv-loader']//li");

    private final FirefoxDriver driver;
    private final WebDriverWait wait;

    public TradeitBot(FirefoxDriver driver) {
        this.driver = driver;
        this.wait = new WebDriverWait(driver, 30);
        this.driver.get(URL);
        this.wait.until(ExpectedConditions.presenceOfElementLocated(By.id("trading-steps")));
    }

    public List<Map<String, Object>> search(String itemName, String game) {
        var gameOption = driver.findElementByXPath("//div[@id='bot-inv-column']//li[@game='" + game + "']");
        if (!gameOption.getAttribute("class").contains("sactive-game")) {
            wait.until(ExpectedConditions.elementToBeClickable(gameOption)).click();
            waitForItems();
        }

        var searchInput = driver.findElementById("ssearch");
        if (!searchInput.getText().equals(itemName)) {
            var botItems = driver.findElements(ITEMS_LOCATOR);
            var lastItemHTML = !botItems.isEmpty() ? botItems.get(botItems.size() - 1).getAttribute("innerHTML") : "";
            searchInput.sendKeys(itemName);
            waitForRefreshedItems(lastItemHTML);
        }

        var updated = false;
        var botItems = driver.findElements(ITEMS_LOCATOR);
        var lastItemHTML = !botItems.isEmpty() ? botItems.get(botItems.size() - 1).getAttribute("innerHTML") : "";
        for (var botItem : botItems) {
            try {
                var multiplier = botItem.findElement(By.xpath(".//div[@class='multiplier']"));
                wait.until(ExpectedConditions.elementToBeClickable(multiplier)).click();
                updated = true;
            } catch (NoSuchElementException ignored) {
            }
        }
        if (updated) {
            waitForRefreshedItems(lastItemHTML);
        }

        var items = driver.findElements(ITEMS_LOCATOR).stream()
                .map(this::convertWebElementToItem)
                .collect(Collectors.toList());

        searchInput.clear();
        return items;
    }

    private void waitForItems() {
        wait.until(webDriver -> {
            var text = webDriver.findElement(By.id("sinv-empty"));
            var elements = webDriver.findElements(ITEMS_LOCATOR);
            log.info("Waiting for items - Message is displayed: " + text.isDisplayed() + " - Items size: " + elements.size());
            return text.isDisplayed() || !elements.isEmpty();
        });
    }

    private void waitForRefreshedItems(String initialState) {
        wait.withTimeout(Duration.ofSeconds(2)).until(webDriver -> {
            var botItems = driver.findElements(ITEMS_LOCATOR);
            var currentState = !botItems.isEmpty() ? botItems.get(botItems.size() - 1).getAttribute("innerHTML") : "";
            var equals = currentState.equals(initialState);
            log.info("Waiting for updated items - Initial state lenght: " + initialState.length() + " - Current state lenght: " + currentState.length());
            return !equals;
        });
    }

    private Map<String, Object> convertWebElementToItem(WebElement element) {
        var floatLocator = By.xpath(".//span[contains(text(), 'Float')]//parent::div");
        var paintLocator = By.xpath(".//span[contains(text(), 'Paint')]//parent::div");
        var priceLocator = By.xpath(".//span[@class='pricetext']");
        var wearLocator = By.xpath(".//div[@class='quality']");
        var stattrakLocator = By.xpath(".//div[@class='stattrak']");
        var lockedLocator = By.xpath(".//div[@class='bot-icon']");

        var item = new HashMap<String, Object>();
        item.put("name", element.getAttribute("data-original-title"));
        item.put("game", element.getAttribute("data-appid"));
        item.put("floatvalue", valueOf(getText(element, floatLocator), Double::valueOf));
        item.put("paint", valueOf(getText(element, paintLocator), Integer::valueOf));
        item.put("price", valueOf(getText(element, priceLocator), Double::valueOf));
        item.put("wear", getText(element, wearLocator));
        item.put("stattrack", getText(element, stattrakLocator) != null);
        item.put("locked", getText(element, lockedLocator));
        return item;
    }

    private String getText(WebElement element, By locator) {
        try {
            var parent = element.findElement(locator);
            var text = parent.getAttribute("textContent").trim();
            var children = parent.findElements(By.xpath("./*"));
            for (var child : children) {
                text = text.replaceFirst(child.getAttribute("textContent"), "").trim();
            }
            return text;
        } catch (NoSuchElementException e) {
            return null;
        }
    }

    private <T> T valueOf(String text, Function<String, T> valueOfFunction) {
        if (text == null) return null;
        text = text.replaceAll("[^\\d.]", "");
        try {
            return valueOfFunction.apply(text);
        } catch (NumberFormatException ignored) {
            return null;
        }
    }
}
